business/istio_certs.go

```diff
@@ -0,0 +1,143 @@
+package business
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/kubernetes"
+	"gopkg.in/yaml.v2"
+)
+
+type IstioCertsService struct {
+	k8s           kubernetes.ClientInterface
+	businessLayer *Layer
+}
+
+type IstioConfig struct {
+	Certificates []struct {
+		SecretName string   `yaml:"secretName"`
+		DNSNames   []string `yaml:"dnsNames"`
+	} `yaml:"certificates"`
+}
+
+type CertInfo struct {
+	SecretName string    `json:"secretName"`
+	DNSNames   []string  `json:"dnsNames"`
+	Issuer     string    `json:"issuer"`
+	Subject    string    `json:"subject"`
+	NotBefore  time.Time `json:"notBefore"`
+	NotAfter   time.Time `json:"notAfter"`
+	Error      error     `json:"error"`
+}
+
+func (ics *IstioCertsService) GetCertsInfo() ([]CertInfo, error) {
+	if certs, err := ics.getChironCertificates(); err == nil {
+		return certs, nil
+	} else if cert, err := ics.getCertificateFromSecret("cacerts"); err == nil {
+		return []CertInfo{cert}, nil
+	} else {
+		cert, err := ics.getCertificateFromSecret("istio-ca-secret")
+		if err != nil {
+			return nil, err
+		}
+		return []CertInfo{cert}, nil
+	}
+}
+
+func (ics *IstioCertsService) getCertificateFromSecret(secretName string) (CertInfo, error) {
+	cfg := config.Get()
+
+	secret, err := ics.k8s.GetSecret(cfg.IstioNamespace, secretName)
+
+	if err != nil {
+		log.Println(err)
+		return CertInfo{}, err
+	}
+	log.Println(string(secret.Data["ca-cert.pem"]))
+	block, _ := pem.Decode(secret.Data["ca-cert.pem"])
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return CertInfo{SecretName: secretName, Error: err}, nil
+	}
+	log.Println(cert.NotBefore)
+	log.Println(cert.NotAfter)
+	return CertInfo{
+		Issuer:     cert.Issuer.String(),
+		Subject:    cert.Subject.String(),
+		NotBefore:  cert.NotBefore,
+		NotAfter:   cert.NotAfter,
+		SecretName: secretName,
+	}, nil
+}
+
+func (ics *IstioCertsService) getChironCertificates() ([]CertInfo, error) {
+	cfg := config.Get()
+
+	istioConfigMap, err := ics.k8s.GetConfigMap(cfg.IstioNamespace, "istio")
```

Comment:
I think there is a method for picking the istio config map, but I might be wrong. Anyway, the magic string should be replaced by the config variable already present in the configmap.

Reply:
I will check, thanks!.

Reply:
> Anyway, the magic string should be replaced by the config variable already present in the configmap.

The config map name is found in our config here - you should use this rather than hardcode "istio" because it isn't always going to be that name: https://github.com/kiali/kiali-operator/blob/master/deploy/kiali/kiali_cr.yaml#L543